libraries/AP_Filesystem/AP_Filesystem_FlashMemory_LittleFS.cpp

```diff
@@ -0,0 +1,1254 @@
+    // in closedir. This is safe here since the lock is recursive; recursion
+    // also lets the thread with the directory open do reads. Hopefully this
+    // will be fixed upstream so we can remove this quirk.
+    fs_sem.take_blocking();
+
+    return result;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+struct dirent *AP_Filesystem_FlashMemory_LittleFS::readdir(void *ptr)
+{
+    FS_CHECK_ALLOWED(nullptr);
+    WITH_SEMAPHORE(fs_sem);
+
+    DirEntry *pair = static_cast<DirEntry*>(ptr);
+    if (!pair) {
+        errno = EINVAL;
+        return nullptr;
+    }
+
+    lfs_info info;
+    int retval = lfs_dir_read(&fs, &pair->dir, &info);
+    if (retval == 0) {
+        /* no more entries */
+        return nullptr;
+    }
+    if (retval < 0) {
+        // failure
+        errno = errno_from_lfs_error(retval);
+        return nullptr;
+    }
+
+    memset(&pair->entry, 0, sizeof(pair->entry));
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+    pair->entry.d_ino = 0;
+    pair->entry.d_seekoff++;
+#endif
+
+    strncpy(pair->entry.d_name, info.name, MIN(strlen(info.name)+1, sizeof(pair->entry.d_name)));
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+    pair->entry.d_namlen = strlen(info.name);
+#endif
+
+    pair->entry.d_type = info.type == LFS_TYPE_DIR ? DT_DIR : DT_REG;
+
+    return &pair->entry;
+}
+#pragma GCC diagnostic pop
+
+int AP_Filesystem_FlashMemory_LittleFS::closedir(void *ptr)
+{
+    FS_CHECK_ALLOWED(-1);
+    WITH_SEMAPHORE(fs_sem);
+
+    DirEntry *pair = static_cast<DirEntry*>(ptr);
+    if (!pair) {
+        errno = EINVAL;
+        return 0;
+    }
+
+    // Before the close, undo the lock we did in opendir so it's released even
+    // if the close fails. We don't undo it above, as the input being nullptr
+    // means we didn't successfully open the directory and lock.
+    fs_sem.give();
+
+    LFS_CHECK(lfs_dir_close(&fs, &pair->dir));
+
+    delete pair;
+
+    return 0;
+}
+
+int64_t AP_Filesystem_FlashMemory_LittleFS::disk_free(const char *path)
+{
+    FS_CHECK_ALLOWED(-1);
+    WITH_SEMAPHORE(fs_sem);
+    ENSURE_MOUNTED();
+
+    lfs_ssize_t alloc_size = lfs_fs_size(&fs);
+    if (alloc_size < 0) {
+        errno = errno_from_lfs_error(alloc_size);
+        return -1;
+    }
+
+    return disk_space(path) - alloc_size;
+}
+
+int64_t AP_Filesystem_FlashMemory_LittleFS::disk_space(const char *path)
+{
+    return fs_cfg.block_count * fs_cfg.block_size;
+}
+
+bool AP_Filesystem_FlashMemory_LittleFS::retry_mount(void)
+{
+    FS_CHECK_ALLOWED(false);
+    WITH_SEMAPHORE(fs_sem);
+
+    if (!dead) {
+        if (!mounted && !mount_filesystem()) {
+            errno = EIO;
+            return false;
+        }
+
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void AP_Filesystem_FlashMemory_LittleFS::unmount(void)
+{
+    WITH_SEMAPHORE(fs_sem);
+
+    if (mounted && !dead) {
+        if (lfs_unmount(&fs) >= 0) {
+            mounted = false;
+        }
```

Comment:
```suggestion
        lfs_unmount(&fs);
        mounted = false;
```

Not sure what to do with the error code, but the unmount can't fail. The filesystem is unmounted after the function is called no matter the error returned.

Reply:
Should we call `free_all_fds()`? Not exactly sure how this works under fatfs, but it seems dangerous either way.

Reply:
Yes. Thanks, fixed